Don't repeat step data in every batch's arguments

seven1m opened the pull request:
This was causing insane memory usage when you have a lot of steps in series.

Before
------

The entire set of jobs was passed into the `on_success` callback as an argument. Since Sidekiq Batches do not get deleted from Redis for a full day, if you start with 10 MB of job data, 10,000 in-series steps, and 1 KB of data in each step, then the total memory usage is something on the order of:

    (10 MB) + (10 MB - 1 KB) + (10 MB - 2 KB) + (10 MB - 3 KB) + ...
    = approximately 50 GB

After
-----

Now, we store the step data in a temporary Redis key we create. Rather than passing the step data into the `on_success` callback, we just pass the temporary Redis key. The `on_success` callback looks up the data in the key, pops an item off the front, then writes the updated step data to the key again. Redis memory usage with this approach (for the same scenario as above: 10,000 in-series steps of 1 KB each) is something on the order of:

    (10 MB) + (10,000 * 72 B)
    ...or:
    (10 MB) + (720 KB)

seven1m committed:
Don't repeat step data in every batch's arguments

This was causing insane memory usage when you have a lot of steps in series.

Before
------

The entire set of jobs was passed into the `on_success` callback as an
argument. Since Sidekiq Batches do not get deleted from Redis for a full
day, if you start with 10 MB of job data, 10,000 in-series steps, and 1
KB of data in each step, then the total memory usage is something on the
order of:

    (10 MB) + (10 MB - 1 KB) + (10 MB - 2 KB) + (10 MB - 3 KB) + ...
    = approximately 50 GB

After
-----

Now, we store the step data in a temporary Redis key we create. Rather
than passing the step data into the `on_success` callback, we just pass
the temporary Redis key. The `on_success` callback looks up the data in
the key, pops an item off the front, then writes the updated step data
to the key again. Redis memory usage with this approach (for the same
scenario as above: 10,000 in-series steps of 1 KB each) is something on
the order of:

    (10 MB) + (10,000 * 72 B)
    ...or:
    (10 MB) + (720 KB)

seven1m committed:
Add backward-compat with old style of arg passing for batch callback

mcfadden reviewed (approved):
This is fantastic. Definitely better than the weird limits we were trying.

seven1m committed:
Fix result_key_index not properly passed when handling legacy args

This didn't seem to affect the result, but it was making some weird keys
in Redis. :-)
